twine/utils.py

```diff
@@ -99,23 +100,40 @@ def get_config(path: str = "~/.pypirc") -> Dict[str, RepositoryConfig]:
     return dict(config)
 
 
+def validate_url(repository_url: str) -> None:
+    """Validate the given url for allowed schemes and components"""
+
+    # Allowed schemes are http and https, based on whether the repository
+    # supports TLS or not, and scheme and host must be present in the URL
+    validator = (
+        rfc3986.validators.Validator()
+        .allow_schemes("http", "https")
+        .require_presence_of("scheme", "host")
+    )
+
+    url = rfc3986.uri_reference(repository_url)
+    validator.validate(url)
+
+
 def get_repository_from_config(
     config_file: str, repository: str, repository_url: Optional[str] = None
 ) -> RepositoryConfig:
     # Get our config from, if provided, command-line values for the
     # repository name and URL, or the .pypirc file
-    if repository_url and "://" in repository_url:
+
+    if repository_url:
+        try:
+            validate_url(repository_url)
+        except rfc3986.exceptions.RFC3986Exception as exc:
+            raise exceptions.UnreachableRepositoryURLDetected(
+                f"Invalid repository URL: {exc.args[0]}."
+            )
```

Comment:
Good call on the `args[0]`; I misunderstood how the exception would be string-ified.

What do you think about moving the `try`/`except` to `validate_url`? That would keep all the `rfc3986`-related code in one place, and further improve the readability of this function.

```suggestion
    if repository_url:
        validate_url(repository_url)
```

Reply:
I have moved the exception inside the validate function in latest commit